### PR TITLE
TRUNK-5146: Use Java streams in UserServiceImpl#checkPrivileges(User)

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/TRUNK-5146
in this ticket , created a subclass checkSuperUserRole and modified it whith checkPrivilege. Added some tests for that
  cc @ibacher 

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by sherrif10** [comment]
Let me resolve the travis failure

**Comment by sherrif10** [comment]
The test is running locally so am not sure where could be the problem, the failure of travis seems to be have come from tests but my tests are locally running cc @ibacher

File: api/src/main/resources/messages.properties [hunk 1]
```diff
@@ -2177,6 +2177,7 @@ User.creating.already.exists=The user {0} already exists, it could therefore not
 User.setPersonId=You need to call setPerson(Person)
 User.you.must.have.privilege=You must have privilege '{0}' in order to assign it.
 User.you.must.have.role=You must have the role '{0}' in order to assign it.
+User.you.must.have.privilege=User you must have the following privileges in order to assign them: {0}
```
Comment (ibacher) [inline api/src/main/resources/messages.properties:2180]: This is the cause of the error. This line should be `User.you.must.have.privileges` (note the "s" on the end).

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 2]
```diff
@@ -280,7 +282,43 @@ public void createUser_shouldNotAllowSystemIdEqualsUsernameWithLuhnCheckDigit()
...
+					User u = new User();
+					u.setPerson(new Person());
+					// assign the specified role to the user
+					u.addName(new PersonName("Benjamin", "A", "Wolfe"));
+					u.setUsername("bwolfe");
+					u.getPerson().setGender("M");
+					u.addRole(role);
+					// here we expect the exception to be thrown
+					userService.createUser(u, "Openmr5xy");
+				});
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java:315]: Could you un-indent these lines here by one tab?

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 3]
```diff
@@ -280,7 +282,43 @@ public void createUser_shouldNotAllowSystemIdEqualsUsernameWithLuhnCheckDigit()
...
+				// we expect this to fail because the currently logged-in user lacks a privilege to be
+				// assigned to the new user
+				expectedException.expect(APIException.class);
+				expectedException.expectMessage("You must have privilege {0} in order to assign it");
+				// set current user to the user defined above
+				withCurrentUserAs(currentUser, () -> {
+					// create a role to assign to the new user
+					Role role = new Role();
+					role.setRole(RoleConstants.AUTHENTICATED);
+					role.addPrivilege(new Privilege("Custom Privilege"));// add a privilege to the role
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java:303]: The comment here should be above the line it comments on

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 4]
```diff
@@ -280,7 +282,43 @@ public void createUser_shouldNotAllowSystemIdEqualsUsernameWithLuhnCheckDigit()
...
+					u.getPerson().setGender("M");
+					u.addRole(role);
+					// here we expect the exception to be thrown
+					userService.createUser(u, "Openmr5xy");
+				});
+	}
+				
+	@Test
+	public void createUser_shouldNotAllowAssigningSuperUserRoleIfCurrentUserDoesnotHaveAssignSystemDeveloperPrivileges() throws IllegalAccessException {
+
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java:320]: The body here should look identical to the above test except that we should assign two privileges to `role` and we should expect the `User you must have the following privileges in order to assign them: {0}` message.

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 2 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java:315]: ok thanks @ibacher  its coming to normalise but i still have a problem with the first test

Comment (ibacher) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java:315]: What sort of problem? It seems to pass in Travis, so if there's an issue, I'd like to get to the bottom of it quickly.

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 4 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java:320]: Am trying to do as you suggested but there is problem with defining method

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 2 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java]: Travis passed this time around, so am remaining with writing tests any hint.

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 4 above)
Comment (ibacher) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java:321]: What kind of problem?

Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java:321]: Everything is okey now, i hope to finish tests on sunday

Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java]: cc.@ibacher, kindly look through this code, thanks

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 3 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java]: have fixed it thanks

**Comment by sherrif10** [comment]
Hello @ibacher  @mks-d  @dkayiwa  kindly review this  PR  guidance is highly appreciated, everything in this context is fine

**Comment by ibacher** [comment]
@sherrif10 I particularly want to make sure we have the test @mks-d recommended earlier before we merge this (i.e. to verify that `checkSuperUserPrivilege()` does what it is expected to do). This is important to ensure we don't have to do an emergency fix. I'm not sure I'm comfortable merging this code without such a test in place.

I was suggesting doing the test with two privileges simply to help you get familiar with what a test for this kind of code might look like. That test should be done, but is less essential than the test for `checkSuperUserPrivilege()`.

**Comment by sherrif10** [comment]
ok thanks sure i will create a test

**Comment by sherrif10** [comment]
cc @ibacher @mks-d @dkayiwa

**Comment by ibacher** [comment]
@sherrif10 So the test failure is because the message in the exception was "You must have the role {0} in order to assign it." and this is, in fact, the message we'd expect when the `checkSuperUserPrivilege()` raises an exception. So if you can modify the `expectedException` in the test, this should be good to go.

**Comment by sherrif10** [comment]
Hey @ibacher   even though i modify the test as it need to be for checkSuperUserPrivilege() the still fails, i dont know if the logic  of checkSuperUserPrivilege() was  the one defined wrongly. am even so confused

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 5]
```diff
@@ -353,7 +350,7 @@ public void createUser_shouldNotAllowAssigningSuperUserRoleIfCurrentUserDoesnotH
 			u.addRole(role);
 			// here we expect the exception to be thrown
 			userService.createUser(u, "Openmr5xy");
-		});
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: This commit does way too much, which is why the test fails.

* The point of wrapping things in the `withCurrentUserAs()` function is that this enables us to set the current executing user to a particular user. By default, the executing user for these tests is "admin", so this test won't fail as expected.
* The call to `role.hasPrivilege(PrivilegeConstants.ASSIGN_SYSTEM_DEVELOPER_ROLE);` doesn't do anything at all. Or rather, it checks to see if the role has that privilege but by itself, that won't throw the error we're expecting, which happens in the `checkSuperUserPrivilege()` code.

My advice would be to undo this last commit and make a much simpler commit that simply changes this line:

```java
expectedException.expectMessage("User you must have the following privileges in order to assign them: {0}"); 
```

To this:

```java
expectedException.expectMessage("You must have the role {0} in order to assign it"); 
```

You can undo the last commit by running: `git reset --hard HEAD~1`.

Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java]: ok then let me do that

**Comment by sherrif10** [comment]
Hey @ibacher , finally all tests are passing kindly take a look through because these tests were too logical but finally they are passing now. Am waiting for your comment on this. Thanks again for your time and persisting on this

**Comment by sherrif10** [comment]
Hey @ibacher am squashing this code though am seeing failing tests anything more

**Comment by ibacher** [comment]
@sherrif10 Don't worry about squashing things. I'll fix things up from here.

**Comment by sherrif10** [comment]
@ibacher  great thanks for it finally , it was a hustle

**Comment by sherrif10** [comment]
Thanks @ibacher ,@mks-d

**Pull request merged**